test/e2e/scenario/recorder/viewports.scenario.ts

```diff
@@ -0,0 +1,371 @@
+>
+`
+
+describe('recorder', () => {
+  beforeEach(() => {
+    if (isGestureUnsupported()) {
+      pending('no touch gesture support')
+    }
+  })
+
+  describe('layout viewport properties', () => {
+    createTest('getWindowWidth/Height should not be affected by pinch zoom')
+      .withRum({ enableExperimentalFeatures: ['visualviewport'] })
+      .withRumInit(initRumAndStartRecording)
+      .withSetup(bundleSetup)
+      .withBody(html`${VIEWPORT_META_TAGS}`)
+      .run(async ({ events }) => {
+        await buildScrollablePage()
+
+        const { innerWidth, innerHeight } = await getWindowInnerDimensions()
+        await performSignificantZoom()
+
+        await browserExecute(() => {
+          window.dispatchEvent(new Event('resize'))
+        })
+
+        const lastViewportResizeData = (
+          await getLastRecord(events, (segment) =>
+            findAllIncrementalSnapshots(segment, IncrementalSource.ViewportResize)
+          )
+        ).data as ViewportResizeData
+
+        const scrollbarWidth = await getScrollbarCorrection()
+
+        expectToBeNearby(lastViewportResizeData.width, innerWidth - scrollbarWidth)
+        expectToBeNearby(lastViewportResizeData.height, innerHeight - scrollbarWidth)
+      })
+
+    /**
+     * window.ScrollX/Y on some devices/browsers are changed by pinch zoom
+     * We need to ensure that our measurements are not affected by pinch zoom
+     */
+    createTest('getScrollX/Y should not be affected by pinch scroll')
+      .withRum({ enableExperimentalFeatures: ['visualviewport'] })
+      .withRumInit(initRumAndStartRecording)
+      .withSetup(bundleSetup)
+      .withBody(html`${VIEWPORT_META_TAGS}`)
+      .run(async ({ events }) => {
+        const VISUAL_SCROLL_DOWN_PX = 60
+        const LAYOUT_SCROLL_AMOUNT = 20
+
+        await buildScrollablePage()
+        await performSignificantZoom()
+        await resetLayoutScroll()
+
+        const initialVisualViewport = await getVisualViewport()
+        const { scrollX: initialScrollX, scrollY: initialScrollY } = await getWindowScroll()
+
+        // Add Visual Viewport Scroll
+        await pinchScrollVerticallyDown(VISUAL_SCROLL_DOWN_PX)
+
+        // Add Layout Viewport Scroll
+        await layoutScrollTo(LAYOUT_SCROLL_AMOUNT, LAYOUT_SCROLL_AMOUNT)
+
+        const nextVisualViewport = await getVisualViewport()
+        const { scrollX: nextScrollX, scrollY: nextScrollY } = await getWindowScroll()
+
+        await browserExecute(() => {
+          document.dispatchEvent(new Event('scroll'))
+        })
+
+        const lastScrollData = (
+          await getLastRecord(events, (segment) => findAllIncrementalSnapshots(segment, IncrementalSource.Scroll))
+        ).data as ScrollData
+
+        // Height changes because URL address bar changes due to scrolling
+        const navBarHeightChange = nextVisualViewport.height - initialVisualViewport.height
+        expect(navBarHeightChange).toBeLessThanOrEqual(30)
```

Comment:
where does 30 come from?

Reply:
The height change is usually be 20px, empirically. So add a bit of a buffer.

Reply:
what about extracting a constant for it?